delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/model/StorageValue.scala

```diff
@@ -65,35 +99,97 @@ object StorageValue {
+      */
+    override def decrypt(
+        crypto: Crypto
+    )(implicit ev: A =:= Encrypted): Either[InvalidEncryptionSecrets, S3StorageValue[Decrypted]] =
+      (accessKey.traverse(crypto.decrypt), secretKey.traverse(crypto.decrypt))
+        .mapN { case (decryptedAccessKey, decryptedSecret) =>
+          copy(accessKey = decryptedAccessKey, secretKey = decryptedSecret, encryptionState = Decrypted)
+        }
+        .leftMap(InvalidEncryptionSecrets(tpe, _))
   }
 
   /**
     * Resolved values to create/update a Remote disk storage
     *
     * @see [[StorageFields.RemoteDiskStorageFields]]
     */
-  final case class RemoteDiskStorageValue(
+  final case class RemoteDiskStorageValue[A <: EncryptionState](
       default: Boolean,
       endpoint: Uri,
-      credentials: Option[AuthToken],
+      credentials: Option[String],
       folder: Label,
       readPermission: Permission,
       writePermission: Permission,
-      maxFileSize: Long
-  ) extends StorageValue {
+      maxFileSize: Long,
+      encryptionState: A
+  ) extends StorageValue[A] {
     override val tpe: StorageType = StorageType.RemoteDiskStorage
+
+    override type This[B <: EncryptionState] = RemoteDiskStorageValue[B]
+
+    /**
+      * Encrypt the credentials of a decrypted [[RemoteDiskStorageValue]]
+      */
+    override def encrypt(
+        crypto: Crypto
+    )(implicit ev: A =:= Decrypted): Either[InvalidEncryptionSecrets, RemoteDiskStorageValue[Encrypted]] =
+      credentials
+        .traverse(crypto.encrypt)
+        .map { encryptedCredentials =>
+          copy(credentials = encryptedCredentials, encryptionState = Encrypted)
+        }
+        .leftMap(InvalidEncryptionSecrets(tpe, _))
+
+    /**
+      * Decrypt the credentials of an encrypted [[RemoteDiskStorageValue]]
+      */
+    override def decrypt(
+        crypto: Crypto
+    )(implicit ev: A =:= Encrypted): Either[InvalidEncryptionSecrets, RemoteDiskStorageValue[Decrypted]] =
+      credentials
+        .traverse(crypto.decrypt)
+        .map { encryptedCredentials =>
+          copy(credentials = encryptedCredentials, encryptionState = Decrypted)
+        }
+        .leftMap(InvalidEncryptionSecrets(tpe, _))
   }
 
   @nowarn("cat=unused")
-  implicit private[model] val storageValueEncoder: Encoder[StorageValue] = {
+  implicit private[model] val storageValueEncoder: Encoder[StorageValue[Decrypted]] = {
     implicit val config: Configuration      = Configuration.default.withDiscriminator(keywords.tpe)
     implicit val pathEncoder: Encoder[Path] = Encoder.encodeString.contramap(_.toString)
 
     Encoder.encodeJsonObject.contramapObject { storage =>
-      deriveConfiguredEncoder[StorageValue].encodeObject(storage).add(keywords.tpe, storage.tpe.iri.asJson)
+      deriveConfiguredEncoder[StorageValue[Decrypted]]
+        .encodeObject(storage)
+        .add(keywords.tpe, storage.tpe.iri.asJson)
+        .remove("credentials")
```

Comment:
maybe make these fields constants to avoid typos

Reply:
I've removed most of the usages with the introduction of the `Secret` class